src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -1639,15 +1662,66 @@ public bool ContainsFile(string fullPath)
             }
 
-            public void ProcessFileChange(string filePath)
-                => ProcessFileChange(filePath, filePath);
+            public async ValueTask ProcessRegularFileChangesAsync(ImmutableArray<string> filePaths)
+            {
+                using (await _project._gate.DisposableWaitAsync().ConfigureAwait(false))
+                {
+                    // If our project has already been removed, this is a stale notification, and we can disregard.
+                    if (_project.HasBeenRemoved)
+                    {
+                        return;
+                    }
+
+                    var documentsToChange = new List<(DocumentId, TextLoader)>(capacity: filePaths.Length);
+
+                    foreach (var filePath in filePaths)
+                    {
+                        if (_documentPathsToDocumentIds.TryGetValue(filePath, out var documentId))
+                        {
+                            // We create file watching prior to pushing the file to the workspace in batching, so it's
+                            // possible we might see a file change notification early. In this case, toss it out. Since
+                            // all adds/removals of documents for this project happen under our lock, it's safe to do this
+                            // check without taking the main workspace lock. We don't have to check for documents removed in
+                            // the batch, since those have already been removed out of _documentPathsToDocumentIds.
+                            if (!_documentsAddedInBatch.Any(d => d.Id == documentId))
+                            {
+                                documentsToChange.Add((documentId, new FileTextLoader(filePath, defaultEncoding: null)));
+                            }
+                        }
+                    }
+
+                    // Nothing actually matched, so we're done
+                    if (documentsToChange.Count == 0)
+                    {
+                        return;
+                    }
+
+                    await _project._workspace.ApplyBatchChangeToWorkspaceMaybeAsync(useAsync: true, s =>
+                    {
+                        var accumulator = new SolutionChangeAccumulator(s);
+
+                        foreach (var (documentId, textLoader) in documentsToChange)
+                        {
+                            if (!s.Workspace.IsDocumentOpen(documentId))
+                            {
+                                accumulator.UpdateSolutionForDocumentAction(
+                                    _documentTextLoaderChangedAction(accumulator.Solution, documentId, textLoader),
+                                    _documentChangedWorkspaceKind,
+                                    SpecializedCollections.SingletonEnumerable(documentId));
+                            }
+                        }
+
+                        return accumulator;
+                    }).ConfigureAwait(true);
```

Comment:
📝 I'm fine with everything in the review except I don't understand how this block came to be. How is this derived from the code in former `ProcessFileChange` method, and why is it correct?

Reply:
So that corresponds to this block: https://github.com/jasonmalinowski/roslyn/blob/153bdd3459619bef2fa3a75c9246fa18c3aa4a62/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs#L1664-L1673

(that link the state after the first commit that split the methods)

This also takes advantage of the general support we have for applying batch operations; the accumulator has logic to say if a single file is changing, raise WorkspaceKind.DocumentChanged; if we're changing multiple files the correct kind to raise is ProjectChanged, so we'll raise tha tinstead.